packages/cinterop/src/commonMain/kotlin/io/realm/internal/interop/RealmInterop.kt

```diff
@@ -46,26 +43,80 @@ expect val INVALID_PROPERTY_KEY: PropertyKey
 @JvmInline
 value class RegistrationToken(val value: Long)
 
+// Pure marker interfaces corresponding to the C-API realm_x_t struct types
+interface CapiT
+interface RealmConfigT : CapiT
+interface RealmSchemaT : CapiT
+interface RealmT : CapiT
+interface LiveRealmT : RealmT
+interface FrozenRealmT : RealmT
+interface RealmObjectT : CapiT
+interface RealmListT : CapiT
+interface RealmResultsT : CapiT
+interface RealmQueryT : CapiT
+interface RealmNotificationTokenT : CapiT
+interface RealmChangesT : CapiT
+interface RealmObjectChangesT : RealmChangesT
+interface RealmCollectionChangesT : RealmChangesT
+// Public type aliases binding to internal verbose type safe type definitions. This should allow us
+// to easily change implementation details later on.
+typealias RealmNativePointer = NativePointer<out CapiT>
+typealias RealmConfigurationPointer = NativePointer<RealmConfigT>
+typealias RealmSchemaPointer = NativePointer<RealmSchemaT>
+typealias RealmPointer = NativePointer<out RealmT>
+typealias LiveRealmPointer = NativePointer<LiveRealmT>
+typealias FrozenRealmPointer = NativePointer<FrozenRealmT>
+typealias RealmObjectPointer = NativePointer<RealmObjectT>
+typealias RealmListPointer = NativePointer<RealmListT>
+typealias RealmResultsPointer = NativePointer<RealmResultsT>
+typealias RealmQueryPointer = NativePointer<RealmQueryT>
+// FIXME Should we differentiate this by generics
```

Comment:
IMO no... typealiasing is for making the types easier to reason about. Adding generics goes in the opposite direction